core/capabilities/webapi/outgoing_connector_handler.go

```diff
@@ -0,0 +1,223 @@
+}
+
+func NewOutgoingConnectorHandler(gc connector.GatewayConnector, config Config, method string, lgger logger.Logger) (*OutgoingConnectorHandler, error) {
+	rateLimiter, err := common.NewRateLimiter(config.RateLimiter)
+	if err != nil {
+		return nil, err
+	}
+
+	// if this is not specified we default to uuid.New
+	// this allows us to have deterministic uuid on tests.
+	if config.UUIDGeneratorFn == nil {
+		config.UUIDGeneratorFn = uuid.New
+	}
+
+	if !validMethod(method) {
+		return nil, fmt.Errorf("invalid outgoing connector handler method: %s", method)
+	}
+
+	responseChs := make(map[string]chan *api.Message)
+	return &OutgoingConnectorHandler{
+		gc:            gc,
+		method:        method,
+		responseChs:   responseChs,
+		responseChsMu: sync.Mutex{},
+		rateLimiter:   rateLimiter,
+		generateUUID:  config.UUIDGeneratorFn,
+		lggr:          lgger,
+	}, nil
+}
+
+// HandleSingleNodeRequest sends a request to first available gateway node and blocks until response is received
+// TODO: handle retries and timeouts
+func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context, messageID string, payload []byte) (*api.Message, error) {
+	ch := make(chan *api.Message, 1)
+	c.responseChsMu.Lock()
+	c.responseChs[messageID] = ch
+	c.responseChsMu.Unlock()
+	l := logger.With(c.lggr, "messageID", messageID)
+	l.Debugw("sending request to gateway")
+
+	body := &api.MessageBody{
+		MessageId: messageID,
+		DonId:     c.gc.DonID(),
+		Method:    c.method,
+		Payload:   payload,
+	}
+
+	// simply, send request to first available gateway node from sorted list
+	// this allows for deterministic selection of gateway node receiver for easier debugging
+	gatewayIDs := c.gc.GatewayIDs()
+	if len(gatewayIDs) == 0 {
+		return nil, errors.New("no gateway nodes available")
+	}
+	sort.Strings(gatewayIDs)
+
+	err := c.gc.SignAndSendToGateway(ctx, gatewayIDs[0], body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to send request to gateway")
+	}
+
+	select {
+	case resp := <-ch:
+		return resp, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (c *OutgoingConnectorHandler) HandleGatewayMessage(ctx context.Context, gatewayID string, msg *api.Message) {
+	body := &msg.Body
+	l := logger.With(c.lggr, "gatewayID", gatewayID, "method", body.Method, "messageID", msg.Body.MessageId)
+	if !c.rateLimiter.Allow(body.Sender) {
+		// error is logged here instead of warning because if a message from gateway is rate-limited,
+		// the workflow will eventually fail with timeout as there are no retries in place yet
+		c.lggr.Errorw("request rate-limited")
+		return
+	}
+	l.Debugw("handling gateway request")
+	switch body.Method {
+	case capabilities.MethodWebAPITarget:
+		var payload capabilities.TargetResponsePayload
+		unmarshalAndSendResponse(ctx, msg, payload, c, l)
+	case capabilities.MethodComputeAction:
+		var payload sdk.FetchResponse
+		unmarshalAndSendResponse(ctx, msg, payload, c, l)
```

Comment:
@agparadiso This is more of a question: what are the differences between FetchResponse and TargetResponsePayload? Why do we need to handle each type here?

Reply:
on chainlink-commons when we marshal the response we use:
```
type FetchResponse struct {
	Success    bool           `json:"success"`           // true if HTTP request was successful
	StatusCode uint8          `json:"statusCode"`        // HTTP status code
	Headers    map[string]any `json:"headers,omitempty"` // HTTP headers
	Body       []byte         `json:"body,omitempty"`    // HTTP response body
}
```

TargetResponsePayload looks different:
```
type TargetResponsePayload struct {
	ExecutionError bool              `json:"executionError"`         // true if there were non-HTTP errors. false if HTTP request was sent regardless of status (2xx, 4xx, 5xx)
	ErrorMessage   string            `json:"errorMessage,omitempty"` // error message in case of failure
	StatusCode     int               `json:"statusCode,omitempty"`   // HTTP status code
	Headers        map[string]string `json:"headers,omitempty"`      // HTTP headers
	Body           []byte            `json:"body,omitempty"`         // HTTP response body
}
```

Reply:
@agparadiso This looks very similar to me; there are just some minor changes to how we deal with success or not -- in fact ExecutionMessage + ExecutionError seem more rich as an API. What would the consequences be of removing using TargetResponsePayload everywhere instead of FetchResponse?

Reply:
Changed and renamed 👍🏼